File: geohash3d.js
/**
 * geohash3d.js — 3D spatial geohash encoder/decoder for Punkto/Punkti protocol
 *
 * Algorithm:
 *   - Base32 alphabet: '0123456789bcdefghjkmnpqrstuvwxyz'
 *   - 12 chars × 5 bits = 60 bits total
 *   - Split as 20 bits lat + 20 bits lon + 20 bits alt
 *   - Interleave order per bit position i (0-indexed from MSB):
 *       i%3 === 0 → lat bit
 *       i%3 === 1 → lon bit
 *       i%3 === 2 → alt bit
 *   - lat range : -90  to  90
 *   - lon range : -180 to  180
 *   - alt range : -500 to  8500  (9000 m total)
 */

const BASE32 = '0123456789bcdefghjkmnpqrstuvwxyz';
const BASE32_MAP = {};
for (let i = 0; i < BASE32.length; i++) BASE32_MAP[BASE32[i]] = i;

const LAT_MIN = -90,  LAT_MAX =  90;
const LON_MIN = -180, LON_MAX =  180;
const ALT_MIN = -500, ALT_MAX =  8500;
const BITS = 20; // bits per dimension
const PRECISION = 12; // chars

/**
 * Normalize a value in [min, max] to an integer in [0, 2^bits - 1]
 */
function normToInt(val, min, max, bits) {
  const range = max - min;
  const norm = (val - min) / range;
  const maxInt = (1 << bits) >>> 0; // 2^bits as unsigned
  // clamp
  const clamped = Math.max(0, Math.min(1 - Number.EPSILON, norm));
  return Math.floor(clamped * maxInt);
}

/**
 * Convert an integer back to coordinate value (center of cell)
 */
function intToVal(intVal, min, max, bits) {
  const maxInt = (1 << bits) >>> 0;
  const cellSize = (max - min) / maxInt;
  return min + (intVal + 0.5) * cellSize;
}

/**
 * Encode lat, lon, alt into a 12-character Base32 3D geohash string.
 * @param {number} lat  latitude  in [-90, 90]
 * @param {number} lon  longitude in [-180, 180]
 * @param {number} alt  altitude  in [-500, 8500]  (metres)
 * @param {number} precision  number of chars (default 12)
 * @returns {string} 12-char Base32 geohash
 */
export function encode(lat, lon, alt = 0, precision = PRECISION) {
  const latInt = normToInt(lat, LAT_MIN, LAT_MAX, BITS);
  const lonInt = normToInt(lon, LON_MIN, LON_MAX, BITS);
  const altInt = normToInt(alt, ALT_MIN, ALT_MAX, BITS);

  // Total bits = precision * 5
  const totalBits = precision * 5;
  // We use 20 bits per dimension; fill MSB-first
  // bit position i (0 = MSB of output):
  //   component = i % 3: 0→lat, 1→lon, 2→alt
  //   within-component bit index:
  //     lat gets bits at positions 0,3,6,...  → latBitIdx = floor(i/3)
  //     lon gets bits at positions 1,4,7,...  → lonBitIdx = floor(i/3)
  //     alt gets bits at positions 2,5,8,...  → altBitIdx = floor(i/3)
  // Since totalBits = 60 and BITS = 20, each dimension gets exactly 20 bits.

  let chars = '';
  let acc = 0;
  let accBits = 0;

  for (let i = 0; i < totalBits; i++) {
    const dim = i % 3; // 0=lat, 1=lon, 2=alt
    const dimBitIdx = Math.floor(i / 3); // 0..19 (MSB first within dimension)
    const bitFromMSB = BITS - 1 - dimBitIdx; // bit position in integer (0=LSB)

    let intVal;
    if (dim === 0) intVal = latInt;
    else if (dim === 1) intVal = lonInt;
    else intVal = altInt;

    const bit = (intVal >>> bitFromMSB) & 1;
    acc = (acc << 1) | bit;
    accBits++;

    if (accBits === 5) {
      chars += BASE32[acc];
      acc = 0;
      accBits = 0;
    }
  }

  return chars;
}

/**
 * Decode a 3D geohash string into {lat, lon, alt, error: {lat, lon, alt}}
 * Returns center of cell + error (half cell size per dimension).
 * @param {string} hash  Base32 geohash string (up to 12 chars)
 * @returns {{lat:number, lon:number, alt:number, error:{lat:number,lon:number,alt:number}}}
 */
export function decode(hash) {
  const bounds = toBounds(hash);
  return {
    lat: (bounds.minLat + bounds.maxLat) / 2,
    lon: (bounds.minLon + bounds.maxLon) / 2,
    alt: (bounds.minAlt + bounds.maxAlt) / 2,
    error: {
      lat: (bounds.maxLat - bounds.minLat) / 2,
      lon: (bounds.maxLon - bounds.minLon) / 2,
      alt: (bounds.maxAlt - bounds.minAlt) / 2,
    },
  };
}

/**
 * Return the bounding box of a 3D geohash.
 * @param {string} hash  Base32 geohash string
 * @returns {{minLat,maxLat,minLon,maxLon,minAlt,maxAlt}}
 */
export function toBounds(hash) {
  const len = hash.length;
  const totalBits = len * 5;

  // Determine how many bits each dimension gets
  // bit positions 0,3,6,... → lat; 1,4,7,... → lon; 2,5,8,... → alt
  let latBits = 0, lonBits = 0, altBits = 0;
  for (let i = 0; i < totalBits; i++) {
    const dim = i % 3;
    if (dim === 0) latBits++;
    else if (dim === 1) lonBits++;
    else altBits++;
  }

  // Reconstruct per-dimension integer values from the hash bits
  let latInt = 0, lonInt = 0, altInt = 0;

  for (let i = 0; i < totalBits; i++) {
    const charIdx = Math.floor(i / 5);
    const bitInChar = 4 - (i % 5); // MSB first within char
    const charVal = BASE32_MAP[hash[charIdx]];
    if (charVal === undefined) throw new Error(`Invalid Base32 character: '${hash[charIdx]}'`);
    const bit = (charVal >>> bitInChar) & 1;

    const dim = i % 3;
    const dimBitIdx = Math.floor(i / 3); // which bit within dimension (MSB first)

    if (dim === 0) {
      latInt = (latInt << 1) | bit;
    } else if (dim === 1) {
      lonInt = (lonInt << 1) | bit;
    } else {
      altInt = (altInt << 1) | bit;
    }
  }

  // Convert integer ranges to coordinate bounds
  const latRange = LAT_MAX - LAT_MIN;
  const lonRange = LON_MAX - LON_MIN;
  const altRange = ALT_MAX - ALT_MIN;

  const latMaxInt = (1 << latBits) >>> 0;
  const lonMaxInt = (1 << lonBits) >>> 0;
  const altMaxInt = (1 << altBits) >>> 0;

  const latCellSize = latRange / latMaxInt;
  const lonCellSize = lonRange / lonMaxInt;
  const altCellSize = altRange / altMaxInt;

  const minLat = LAT_MIN + latInt * latCellSize;
  const minLon = LON_MIN + lonInt * lonCellSize;
  const minAlt = ALT_MIN + altInt * altCellSize;

  return {
    minLat,
    maxLat: minLat + latCellSize,
    minLon,
    maxLon: minLon + lonCellSize,
    minAlt,
    maxAlt: minAlt + altCellSize,
  };
}

// ---------------------------------------------------------------------------
// Self-test (uncomment to run in browser console or Node.js)
// ---------------------------------------------------------------------------
// function selfTest() {
//   const tests = [
//     { lat: 55.6761,  lon: 12.5683,  alt: 10,   desc: 'Copenhagen' },
//     { lat: 51.5074,  lon: -0.1278,  alt: 11,   desc: 'London' },
//     { lat: 40.7128,  lon: -74.0060, alt: 10,   desc: 'New York' },
//     { lat: 35.6762,  lon: 139.6503, alt: 40,   desc: 'Tokyo' },
//     { lat: -33.8688, lon: 151.2093, alt: 50,   desc: 'Sydney' },
//     { lat: 0,        lon: 0,        alt: 0,    desc: 'Null Island' },
//     { lat: -90,      lon: -180,     alt: -500, desc: 'Min corner' },
//     { lat: 89.9999,  lon: 179.9999, alt: 8499, desc: 'Near max corner' },
//     { lat: 55.6761,  lon: 12.5683,  alt: 0,    desc: 'Copenhagen alt=0' },
//   ];
//
//   console.log('=== geohash3d self-test ===');
//   let pass = 0, fail = 0;
//
//   for (const t of tests) {
//     const hash = encode(t.lat, t.lon, t.alt);
//     const result = decode(hash);
//     const latOk = Math.abs(result.lat - t.lat) <= result.error.lat;
//     const lonOk = Math.abs(result.lon - t.lon) <= result.error.lon;
//     const altOk = Math.abs(result.alt - t.alt) <= result.error.alt;
//     const ok = latOk && lonOk && altOk;
//     if (ok) pass++; else fail++;
//     console.log(
//       `${ok ? 'PASS' : 'FAIL'} ${t.desc}: hash=${hash}`,
//       `decoded=(${result.lat.toFixed(4)}, ${result.lon.toFixed(4)}, ${result.alt.toFixed(1)})`,
//       `err=(±${result.error.lat.toFixed(4)}, ±${result.error.lon.toFixed(4)}, ±${result.error.alt.toFixed(1)})`,
//     );
//   }
//
//   // Round-trip precision test at 12 chars
//   const h12 = encode(55.6761, 12.5683, 42);
//   const d12 = decode(h12);
//   console.log(`\n12-char precision: lat_err=±${d12.error.lat.toFixed(6)}°, lon_err=±${d12.error.lon.toFixed(6)}°, alt_err=±${d12.error.alt.toFixed(2)}m`);
//
//   console.log(`\nResults: ${pass} passed, ${fail} failed`);
// }
// selfTest();

File: core/geohash3d.py
"""
geohash3d.py — 3D spatial geohash encoder/decoder for Punkto/Punkti protocol

Direct Python port of geohash3d.js.

Algorithm:
  - Base32 alphabet: '0123456789bcdefghjkmnpqrstuvwxyz'
  - 12 chars × 5 bits = 60 bits total
  - Split as 20 bits lat + 20 bits lon + 20 bits alt
  - Interleave order per bit position i (0-indexed from MSB):
      i%3 == 0 → lat bit
      i%3 == 1 → lon bit
      i%3 == 2 → alt bit
  - lat range : -90   to  90
  - lon range : -180  to  180
  - alt range : -500  to  8500  (9000 m total)
"""

from __future__ import annotations
from dataclasses import dataclass

BASE32 = '0123456789bcdefghjkmnpqrstuvwxyz'
BASE32_MAP = {c: i for i, c in enumerate(BASE32)}

LAT_MIN, LAT_MAX = -90.0, 90.0
LON_MIN, LON_MAX = -180.0, 180.0
ALT_MIN, ALT_MAX = -500.0, 8500.0
BITS = 20          # bits per dimension
PRECISION = 12     # default char length


@dataclass
class Bounds3D:
    min_lat: float
    max_lat: float
    min_lon: float
    max_lon: float
    min_alt: float
    max_alt: float


@dataclass
class Decoded3D:
    lat: float
    lon: float
    alt: float
    error_lat: float
    error_lon: float
    error_alt: float


def _norm_to_int(val: float, min_: float, max_: float, bits: int) -> int:
    """Normalize a value in [min_, max_] to an integer in [0, 2^bits - 1]."""
    range_ = max_ - min_
    norm = (val - min_) / range_
    max_int = 1 << bits
    clamped = max(0.0, min(1.0 - 1e-15, norm))
    return int(clamped * max_int)


def encode(lat: float, lon: float, alt: float = 0.0, precision: int = PRECISION) -> str:
    """
    Encode lat, lon, alt into a Base32 3D geohash string.

    Args:
        lat:       latitude  in [-90, 90]
        lon:       longitude in [-180, 180]
        alt:       altitude  in [-500, 8500] metres (default 0)
        precision: number of output characters (default 12)

    Returns:
        Base32 geohash string of length `precision`
    """
    lat_int = _norm_to_int(lat, LAT_MIN, LAT_MAX, BITS)
    lon_int = _norm_to_int(lon, LON_MIN, LON_MAX, BITS)
    alt_int = _norm_to_int(alt, ALT_MIN, ALT_MAX, BITS)

    total_bits = precision * 5
    chars = []
    acc = 0
    acc_bits = 0

    for i in range(total_bits):
        dim = i % 3          # 0=lat, 1=lon, 2=alt
        dim_bit_idx = i // 3 # 0..19, MSB first within dimension
        bit_from_msb = BITS - 1 - dim_bit_idx

        if dim == 0:
            int_val = lat_int
        elif dim == 1:
            int_val = lon_int
        else:
            int_val = alt_int

        bit = (int_val >> bit_from_msb) & 1
        acc = (acc << 1) | bit
        acc_bits += 1

        if acc_bits == 5:
            chars.append(BASE32[acc])
            acc = 0
            acc_bits = 0

    return ''.join(chars)


def to_bounds(hash_str: str) -> Bounds3D:
    """
    Return the 3D bounding box of a geohash string.

    Args:
        hash_str: Base32 geohash string (1–12 chars)

    Returns:
        Bounds3D with min/max lat, lon, alt
    """
    length = len(hash_str)
    total_bits = length * 5

    lat_bits = lon_bits = alt_bits = 0
    for i in range(total_bits):
        dim = i % 3
        if dim == 0:
            lat_bits += 1
        elif dim == 1:
            lon_bits += 1
        else:
            alt_bits += 1

    lat_int = lon_int = alt_int = 0

    for i in range(total_bits):
        char_idx = i // 5
        bit_in_char = 4 - (i % 5)  # MSB first within char
        char_val = BASE32_MAP.get(hash_str[char_idx])
        if char_val is None:
            raise ValueError(f"Invalid Base32 character: {hash_str[char_idx]!r}")
        bit = (char_val >> bit_in_char) & 1
        dim = i % 3
        if dim == 0:
            lat_int = (lat_int << 1) | bit
        elif dim == 1:
            lon_int = (lon_int << 1) | bit
        else:
            alt_int = (alt_int << 1) | bit

    lat_max_int = 1 << lat_bits
    lon_max_int = 1 << lon_bits
    alt_max_int = 1 << alt_bits

    lat_range = LAT_MAX - LAT_MIN
    lon_range = LON_MAX - LON_MIN
    alt_range = ALT_MAX - ALT_MIN

    lat_cell = lat_range / lat_max_int
    lon_cell = lon_range / lon_max_int
    alt_cell = alt_range / alt_max_int

    min_lat = LAT_MIN + lat_int * lat_cell
    min_lon = LON_MIN + lon_int * lon_cell
    min_alt = ALT_MIN + alt_int * alt_cell

    return Bounds3D(
        min_lat=min_lat,
        max_lat=min_lat + lat_cell,
        min_lon=min_lon,
        max_lon=min_lon + lon_cell,
        min_alt=min_alt,
        max_alt=min_alt + alt_cell,
    )


def decode(hash_str: str) -> Decoded3D:
    """
    Decode a 3D geohash string into coordinates (center of cell) + errors.

    Args:
        hash_str: Base32 geohash string (1–12 chars)

    Returns:
        Decoded3D with lat, lon, alt and half-cell error per dimension
    """
    b = to_bounds(hash_str)
    return Decoded3D(
        lat=(b.min_lat + b.max_lat) / 2,
        lon=(b.min_lon + b.max_lon) / 2,
        alt=(b.min_alt + b.max_alt) / 2,
        error_lat=(b.max_lat - b.min_lat) / 2,
        error_lon=(b.max_lon - b.min_lon) / 2,
        error_alt=(b.max_alt - b.min_alt) / 2,
    )
